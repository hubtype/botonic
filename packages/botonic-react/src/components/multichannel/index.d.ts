import * as React from 'react'

type IndexMode = 'number' | 'letter' | undefined //undefined means no index

export interface MultichannelViewOptions {
  boldIndex?: boolean
  indexSeparator?: string // after it, a space will always be added
}

export interface MultichannelContextType extends MultichannelViewOptions {
  currentIndex: number | string // can be letter or number
  /** @see same field at MultichannelProps */
  messageSeparator?: string
}

// Text
export interface MultichannelTextProps extends MultichannelViewOptions {
  index?: string
  indexMode?: IndexMode
  /** Defaults to no separator between lines*/
  newline?: string
}

export const MultichannelText: React.FunctionComponent<MultichannelTextProps>

// Carousel
export interface MultichannelCarouselProps extends MultichannelViewOptions {
  enableURL?: boolean
  indexMode?: IndexMode
  showTitle?: boolean
  showSubtitle?: boolean
}
export const MultichannelCarousel: React.FunctionComponent<MultichannelCarouselProps>

// Button
export interface MultichannelButtonProps {
  newline?: string
}
export const MultichannelButton: React.FunctionComponent<MultichannelButtonProps>

export interface MultichannelProps extends MultichannelViewOptions {
  firstIndex?: number | string
  carousel?: MultichannelCarouselProps
  text?: MultichannelTextProps
  /**
   * If undefined, each component will be served in a different botonic message
   * '' will concatenate in the same line
   * '\n' will split in different lines
   * '\n\n' will separate with an empty line
   **/
  messageSeparator?: string
}

export const Multichannel: React.FunctionComponent<MultichannelProps>
