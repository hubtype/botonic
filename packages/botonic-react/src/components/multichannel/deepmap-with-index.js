import { Children, cloneElement, isValidElement } from 'react'

import { hasComplexChildren } from '../../util/react'

export const deepMapWithIndex = (children, deepMapFn) => {
  return Children.toArray(children).map((child, index) => {
    if (isValidElement(child) && hasComplexChildren(child)) {
      // Clone the child that has children and map them too
      return deepMapFn(
        cloneElement(
          child,
          Object.assign(Object.assign({}, child.props), {
            children: deepMapWithIndex(child.props.children, deepMapFn),
          })
        ),
        index,
        children
      )
    }
    return deepMapFn(child, index, children)
  })
}
