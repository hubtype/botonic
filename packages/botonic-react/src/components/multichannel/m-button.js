import React from 'react'
import { RequestContext } from '../../contexts'
import { Button } from '../button'
import { Providers } from '@botonic/core'

export class MultichannelButton extends React.Component {
  static contextType = RequestContext
  constructor(props) {
    super(props)
  }
  render() {
    if (this.context.session.user.provider == Providers.Messaging.WHATSAPPNEW) {
      if (this.props.url) {
        return `${this.props.children}: ${this.props.url}`
      }
      if (this.props.payload || this.props.path) {
        return `${this.props.children}`
      }
      if (this.props.webview) {
        return <Button {...this.props}>{`${this.props.children}`}</Button>
      }
    } else {
      return <Button {...this.props}>{this.props.children}</Button>
    }
  }
}
