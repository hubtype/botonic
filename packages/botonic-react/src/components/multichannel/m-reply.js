import React from 'react'
import { RequestContext } from '../../contexts'
import { Reply } from '../reply'
import { Providers } from '@botonic/core'

export class MultichannelReply extends React.Component {
  static contextType = RequestContext
  constructor(props) {
    super(props)
  }
  render() {
    if (this.context.session.user.provider == Providers.Messaging.WHATSAPPNEW) {
      if (this.props.payload || this.props.path) {
        return `${this.props.children}`
      }
    } else {
      return <Reply {...this.props}>{this.props.children}</Reply>
    }
  }
}
