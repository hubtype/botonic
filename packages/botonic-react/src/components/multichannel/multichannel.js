import React from 'react'
import { RequestContext } from '../../contexts'

export class Multichannel extends React.Component {
  static contextType = RequestContext
  constructor(props) {
    super(props)
  }
  render() {
    if (this.context.session.user.provider == this.props.channel) {
      return this.props.children
    }
    return null
  }
}
