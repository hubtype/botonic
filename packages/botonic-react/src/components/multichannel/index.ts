import { Multichannel } from './multichannel'
import { MultichannelButton } from './multichannel-button'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelReply } from './multichannel-reply'
import { MultichannelText } from './multichannel-text'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
