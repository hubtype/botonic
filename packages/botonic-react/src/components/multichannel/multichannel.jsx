import React, { useContext } from 'react'
import { RequestContext } from '../../contexts'
import { Text } from '../text'
import { isWhatsapp } from './multichannel-utils'
import { MultichannelContext } from './multichannel-context'
import { MultichannelButton } from './multichannel-button'
import { MultichannelText } from './multichannel-text'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelReply } from './multichannel-reply'
import { deepMapWithIndex } from './deepmap-with-index'

export const Multichannel = props => {
  const requestContext = useContext(RequestContext)

  if (!isWhatsapp(requestContext)) {
    return props.children
  }

  let newChildren = deepMapWithIndex(props.children, (child, index) => {
    if (child && child.type && child.type.name === 'Button') {
      return (
        <MultichannelButton {...child.props}>
          {child.props.children}
        </MultichannelButton>
      )
    }
    if (child && child.type && child.type.name === 'Reply') {
      return (
        <MultichannelReply {...child.props}>
          {child.props.children}
        </MultichannelReply>
      )
    }
    if (child && child.type && child.type.name === 'Text') {
      return (
        <MultichannelText
          {...child.props}
          {...props.text}
          key={props.key}
          {...(props.messageSeparator &&
            index > 0 && { newline: props.messageSeparator })}
        >
          {child.props.children}
        </MultichannelText>
      )
    }
    if (child && child.type && child.type.name === 'Carousel') {
      return (
        <MultichannelCarousel
          {...child.props}
          {...props.carousel}
          key={props.key}
        >
          {child.props.children}
        </MultichannelCarousel>
      )
    }
    return child
  })
  if (props.messageSeparator != null) {
    newChildren = newChildren.map((c, index) =>
      index > 0 && typeof c === 'string' ? props.messageSeparator + c : c
    )
    newChildren = (
      <Text {...props} key={props.key}>
        {newChildren}
      </Text>
    )
  }
  return (
    <MultichannelContext.Provider
      value={{
        currentIndex: props.firstIndex,
        boldIndex: props.boldIndex,
        indexSeparator: props.indexSeparator,
        messageSeparator: props.messageSeparator,
      }}
    >
      {newChildren}
    </MultichannelContext.Provider>
  )
}
