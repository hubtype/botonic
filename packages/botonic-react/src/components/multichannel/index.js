import { Multichannel } from './multichannel'
import { MultichannelText } from './m-text'
import { MultichannelButton } from './m-button'
import { MultichannelReply } from './m-reply'
import { MultichannelCarousel } from './m-carousel'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
