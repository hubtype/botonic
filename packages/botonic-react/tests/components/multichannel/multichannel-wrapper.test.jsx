import {
  Text,
  Button,
  Reply,
  Carousel,
  Element,
  Pic,
  Title,
  Subtitle,
} from '../../../src'
import React from 'react'
import { whatsappRenderer } from '../../helpers/test-utils'
import { Multichannel } from '../../../src/components/multichannel/multichannel'

const LEGACY_PROPS = {
  carousel: {
    indexMode: 'number',
    showTitle: true,
    showSubtitle: true,
  },
  text: {
    indexMode: 'number',
  },
  indexSeparator: '.',
  messageSeparator: null,
}

describe('Multichannel wrapper', () => {
  test('just text', () => {
    const sut = (
      <Multichannel {...LEGACY_PROPS}>
        <Text>Some text</Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('text with buttons', () => {
    const sut = (
      <Multichannel {...LEGACY_PROPS}>
        <Text>
          Some with buttons
          {[
            <Button key={'1'} payload='payload1'>
              Button 1
            </Button>,
            <Button key={'2'} path='path1'>
              Button 2
            </Button>,
            <Button key={'3'} url='http://testurl.com'>
              Visit website
            </Button>,
          ]}
        </Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('text with replies', () => {
    const sut = (
      <Multichannel {...LEGACY_PROPS}>
        <Text>
          Some text with replies
          {[
            <Reply key={'1'} payload='payload1'>
              Reply 1
            </Reply>,
            <Reply key={'2'} path='path1'>
              Reply 2
            </Reply>,
          ]}
        </Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  const carouselWithButtons = [
    {
      title: 'Title1',
      subtitle: 'Subtitle1',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
      buttons: [{ payload: 'Payload1', text: 'Previo a la compra' }],
    },
    {
      title: 'Title2',
      subtitle: 'Subtitle2',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
      buttons: [
        { payload: 'Payload2.1', text: 'Durante la compra' },
        { payload: 'Payload2.2', text: 'Durante la compra2' },
      ],
    },
    {
      title: 'Title3',
      subtitle: 'Subtitle3',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
      buttons: [{ payload: 'Payload3', text: 'Posterior a la compra' }],
    },
  ]
  test('text and carousel legacy', () => {
    const sut = (
      <Multichannel {...LEGACY_PROPS}>
        <Text>This is a multichannel Carousel</Text>

        <Carousel>
          {carouselWithButtons.map((e, i) => (
            <Element key={i}>
              <Pic src={e.pic} />
              <Title>{e.title}</Title>
              <Subtitle>{e.subtitle}</Subtitle>
              {e.buttons.map((e, i) => (
                <Button key={`${e.text}${i}`} payload={e.payload}>
                  {e.text}
                </Button>
              ))}
            </Element>
          ))}
        </Carousel>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('text and carousel only show button text, 1 single message', () => {
    const sut = (
      <Multichannel
        indexSeparator={'.'}
        carousel={{
          indexMode: 'letter',
          showTitle: false,
          showSubtitle: false,
        }}
        messageSeparator={'\n'}
      >
        <Text>This is a multichannel Carousel</Text>

        <Carousel>
          <Element key={0}>
            <Subtitle>Subtitle will not appear</Subtitle>
            {[
              <Button key={1} payload='Payload1'>
                {'Previo'}
              </Button>,
            ]}
          </Element>
          <Element key={1}>
            {[
              <Button key={2} payload='Payload2'>
                {'Durante'}
              </Button>,
            ]}
          </Element>
        </Carousel>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('2 text in 1 single message', () => {
    const sut = (
      <Multichannel messageSeparator={'\n'}>
        <Text>Text1</Text>
        <Text>Text2</Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('1 text in 1 single message', () => {
    const sut = (
      <Multichannel messageSeparator={'\n'}>
        <Text>Text1</Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('2 texts with buttons', () => {
    const sut = (
      <Multichannel messageSeparator={'\n'}>
        <Text>
          Text1
          <Button key={'1'} payload='payload1'>
            Button 1
          </Button>
        </Text>
        <Text>
          Text2
          <Button key={'2'} payload='payload2'>
            Button 2
          </Button>
        </Text>
      </Multichannel>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('many multichannels', () => {
    const sut = (
      <>
        <Multichannel {...LEGACY_PROPS}>
          <Text>
            Some text with replies
            {[
              <Reply key={'1'} payload='payload1'>
                Reply 1
              </Reply>,
              <Reply key={'2'} path='path1'>
                Reply 2
              </Reply>,
            ]}
          </Text>
          <Text>
            Some text with replies
            {[
              <Reply key={'1'} payload='payload1'>
                Reply 1 should have 3.
              </Reply>,
              <Reply key={'2'} path='path1'>
                Reply 2 should have 4.
              </Reply>,
            ]}
          </Text>
        </Multichannel>
        <Multichannel {...LEGACY_PROPS}>
          <Text>
            Some text with replies
            {[
              <Reply key={'1'} payload='payload1'>
                Reply 1 number reset
              </Reply>,
              <Reply key={'2'} path='path1'>
                Reply 2 number reset
              </Reply>,
            ]}
          </Text>
        </Multichannel>
      </>
    )

    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
