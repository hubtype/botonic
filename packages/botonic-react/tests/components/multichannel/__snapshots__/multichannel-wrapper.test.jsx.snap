// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multichannel wrapper 1 text in 1 single message 1`] = `
<message
  delay={0}
  messageSeparator="
"
  type="text"
  typing={0}
>
  Text1
</message>
`;

exports[`Multichannel wrapper 2 text in 1 single message 1`] = `
<message
  delay={0}
  messageSeparator="
"
  type="text"
  typing={0}
>
  Text1
  
Text2
</message>
`;

exports[`Multichannel wrapper 2 texts with buttons 1`] = `
<message
  delay={0}
  messageSeparator="
"
  type="text"
  typing={0}
>
  Text1
  
Button 1
  
Text2
  
Button 2
</message>
`;

exports[`Multichannel wrapper just text 1`] = `
<message
  delay={0}
  indexMode="number"
  type="text"
  typing={0}
>
  Some text
</message>
`;

exports[`Multichannel wrapper many multichannels 1`] = `
Array [
  <message
    delay={0}
    indexMode="number"
    type="text"
    typing={0}
  >
    Some text with replies
    
1. Reply 1
    
2. Reply 2
  </message>,
  <message
    delay={0}
    indexMode="number"
    type="text"
    typing={0}
  >
    Some text with replies
    
3. Reply 1 should have 3.
    
4. Reply 2 should have 4.
  </message>,
  <message
    delay={0}
    indexMode="number"
    type="text"
    typing={0}
  >
    Some text with replies
    
1. Reply 1 number reset
    
2. Reply 2 number reset
  </message>,
]
`;

exports[`Multichannel wrapper text and carousel legacy 1`] = `
Array [
  <message
    delay={0}
    indexMode="number"
    type="text"
    typing={0}
  >
    This is a multichannel Carousel
  </message>,
  <message
    delay={0}
    indexMode="number"
    newkey={0}
    type="text"
    typing={0}
  >
    *Title1* _Subtitle1_
    
1. Previo a la compra
  </message>,
  <message
    delay={0}
    indexMode="number"
    newkey={1}
    type="text"
    typing={0}
  >
    *Title2* _Subtitle2_
    
2. Durante la compra
    
3. Durante la compra2
  </message>,
  <message
    delay={0}
    indexMode="number"
    newkey={2}
    type="text"
    typing={0}
  >
    *Title3* _Subtitle3_
    
4. Posterior a la compra
  </message>,
]
`;

exports[`Multichannel wrapper text and carousel only show button text, 1 single message 1`] = `
<message
  carousel={
    Object {
      "indexMode": "letter",
      "showSubtitle": false,
      "showTitle": false,
    }
  }
  delay={0}
  indexSeparator="."
  messageSeparator="
"
  type="text"
  typing={0}
>
  This is a multichannel Carousel
  
a. Previo
  
b. Durante
</message>
`;

exports[`Multichannel wrapper text with buttons 1`] = `
<message
  delay={0}
  indexMode="number"
  type="text"
  typing={0}
>
  Some with buttons
  
1. Button 1
  
2. Button 2
  
- Visit website: http://testurl.com
</message>
`;

exports[`Multichannel wrapper text with replies 1`] = `
<message
  delay={0}
  indexMode="number"
  type="text"
  typing={0}
>
  Some text with replies
  
1. Reply 1
  
2. Reply 2
</message>
`;
